audioipc/src/sys/unix/mod.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright © 2021 Mozilla Foundation
+//
+// This program is made available under an ISC-style license.  See the
+// accompanying file LICENSE for details
+
+use std::io::Result;
+use std::os::unix::prelude::{AsRawFd, FromRawFd, RawFd};
+
+use bytes::{BufMut, BytesMut};
+use iovec::IoVec;
+use mio::net::UnixStream;
+
+use crate::{close_platform_handle, cmsg, msg, PlatformHandle};
+
+use super::{RecvMsg, SendMsg};
+
+pub struct Pipe(pub UnixStream);
+
+// Create a connected "pipe" pair.  The `Pipe` is the server end,
+// the `PlatformHandle` is the client end to be remoted.
+pub fn make_pipe_pair() -> Result<(Pipe, PlatformHandle)> {
+    let (server, client) = UnixStream::pair()?;
+    Ok((Pipe(server), PlatformHandle::from(client)))
+}
+
+impl Pipe {
+    #[allow(clippy::missing_safety_doc)]
+    pub unsafe fn from_raw_handle(raw: crate::PlatformHandleType) -> Pipe {
+        Pipe(UnixStream::from_raw_fd(raw))
+    }
+}
+
+impl RecvMsg for Pipe {
+    // Receive data (and fds) from the associated connection.  `recv_msg` expects the capacity of
+    // the `ConnectionBuffer` members has been adjusted appropriate by the caller.
+    fn recv_msg(&mut self, buf: &mut ConnectionBuffer) -> Result<usize> {
+        assert!(buf.buf.remaining_mut() > 0);
+        assert!(buf.cmsg.remaining_mut() > 0);
+        #[cfg(target_os = "linux")]
+        let flags = libc::MSG_CMSG_CLOEXEC;
+        #[cfg(not(target_os = "linux"))]
+        let flags = 0;
+        let r = unsafe {
+            let mut iovec = [<&mut IoVec>::from(buf.buf.bytes_mut())];
+            msg::recv_msg_with_flags(self.0.as_raw_fd(), &mut iovec, buf.cmsg.bytes_mut(), flags)
+        };
+        match r {
+            Ok((n, cmsg_n, msg_flags)) => unsafe {
+                debug_assert_eq!(msg_flags, flags);
+                buf.buf.advance_mut(n);
+                buf.cmsg.advance_mut(cmsg_n);
+                Ok(n)
+            },
+            Err(e) => Err(e),
+        }
+    }
+}
+
+impl SendMsg for Pipe {
+    // Send data (and fds) on the associated connection.  `send_msg` adjusts the length of the
+    // `ConnectionBuffer` members based on the size of the successful send operation.
+    fn send_msg(&mut self, buf: &mut ConnectionBuffer) -> Result<usize> {
+        assert!(!buf.buf.is_empty());
+        let r = {
+            let iovec = [<&IoVec>::from(&buf.buf[..buf.buf.len()])];
+            msg::send_msg_with_flags(self.0.as_raw_fd(), &iovec, &buf.cmsg[..buf.cmsg.len()], 0)
+        };
+        match r {
+            Ok(n) => {
+                buf.buf.advance(n);
+                // Close sent fds.
+                let b = buf.cmsg.clone().freeze();
```

Comment:
`b.is_empty` is true for any messages without an associated handle being sent.  Since this layer doesn't know anything about messages, it needs to handle both cases.  When `b.is_empty` is true, `cmsg::iterator` returns `None`.